xbatcher/generators.py

```diff
@@ -1,10 +1,262 @@
 """Classes for iterating through xarray datarrays / datasets in batches."""
 
 import itertools
+import warnings
+from operator import itemgetter
 from typing import Any, Dict, Hashable, Iterator, List, Sequence, Union
 
+import numpy as np
 import xarray as xr
 
+PatchGenerator = Iterator[Dict[Hashable, slice]]
+BatchSelector = List[Dict[Hashable, slice]]
+BatchSelectorSet = Dict[int, BatchSelector]
+
+
+class BatchSchema:
+    """
+    A representation of the indices and stacking/transposing parameters needed
+    to generator batches from Xarray Datasets and DataArrays using
+    xbatcher.BatchGenerator.
+
+    Parameters
+    ----------
+    ds : ``xarray.Dataset`` or ``xarray.DataArray``
+        The data to iterate over. Unlike for the BatchGenerator, the data is
+        not retained as a class attribute for the BatchSchema.
+    input_dims : dict
+        A dictionary specifying the size of the inputs in each dimension,
+        e.g. ``{'lat': 30, 'lon': 30}``
+        These are the dimensions the ML library will see. All other dimensions
+        will be stacked into one dimension called ``sample``.
+    input_overlap : dict, optional
+        A dictionary specifying the overlap along each dimension
+        e.g. ``{'lat': 3, 'lon': 3}``
+    batch_dims : dict, optional
+        A dictionary specifying the size of the batch along each dimension
+        e.g. ``{'time': 10}``. These will always be iterated over.
+    concat_input_dims : bool, optional
+        If ``True``, the dimension chunks specified in ``input_dims`` will be
+        concatenated and stacked into the ``sample`` dimension. The batch index
+        will be included as a new level ``input_batch`` in the ``sample``
+        coordinate.
+        If ``False``, the dimension chunks specified in ``input_dims`` will be
+        iterated over.
+    preload_batch : bool, optional
+        If ``True``, each batch will be loaded into memory before reshaping /
+        processing, triggering any dask arrays to be computed.
+
+    Notes
+    -----
+    The BatchSchema is experimental and subject to change without notice.
+    """
+
+    def __init__(
+        self,
+        ds: Union[xr.Dataset, xr.DataArray],
+        input_dims: Dict[Hashable, int],
+        input_overlap: Dict[Hashable, int] = {},
+        batch_dims: Dict[Hashable, int] = {},
+        concat_input_bins: bool = True,
+        preload_batch: bool = True,
+    ):
```

Comment:
Set `input_overlap=None` and `batch_dims=None` in function definition so as to avoid any potential issues with mutability...

```suggestion
    def __init__(
        self,
        ds: Union[xr.Dataset, xr.DataArray],
        input_dims: Dict[Hashable, int],
        input_overlap: Dict[Hashable, int] = None,
        batch_dims: Dict[Hashable, int] = None,
        concat_input_bins: bool = True,
        preload_batch: bool = True,
    ):
        if input_overlap is None:
            input_overlap = {}
        if batch_dims is None:
            batch_dims = {}
```